File: src/snippets/graph/dijkstra/dijkstra.cs
using System;
using System.Collections.Generic;

public class Graph
{
    private readonly int _vertices;
    private readonly List<Tuple<int, int>>[] _adjacencyList;

    public Graph(int vertices)
    {
        _vertices = vertices;
        _adjacencyList = new List<Tuple<int, int>>[vertices];

        for (int i = 0; i < vertices; i++)
        {
            _adjacencyList[i] = new List<Tuple<int, int>>();
        }
    }

    public void AddEdge(int u, int v, int weight)
    {
        _adjacencyList[u].Add(new Tuple<int, int>(v, weight));
        _adjacencyList[v].Add(new Tuple<int, int>(u, weight)); // For undirected graph
    }

    public int[] Dijkstra(int source)
    {
        int[] distances = new int[_vertices];
        bool[] shortestPathTreeSet = new bool[_vertices];

        for (int i = 0; i < _vertices; i++)
        {
            distances[i] = int.MaxValue;
            shortestPathTreeSet[i] = false;
        }

        distances[source] = 0;

        for (int count = 0; count < _vertices - 1; count++)
        {
            int u = MinDistance(distances, shortestPathTreeSet);
            shortestPathTreeSet[u] = true;

            for (int v = 0; v < _vertices; v++)
            {
                if (!shortestPathTreeSet[v] && _adjacencyList[u].Exists(x => x.Item1 == v) &&
                    distances[u] != int.MaxValue && distances[u] + _adjacencyList[u].Find(x => x.Item1 == v).Item2 < distances[v])
                {
                    distances[v] = distances[u] + _adjacencyList[u].Find(x => x.Item1 == v).Item2;
                }
            }
        }

        return distances;
    }

    private int MinDistance(int[] distances, bool[] shortestPathTreeSet)
    {
        int min = int.MaxValue, minIndex = -1;

        for (int v = 0; v < _vertices; v++)
        {
            if (shortestPathTreeSet[v] == false && distances[v] <= min)
            {
                min = distances[v];
                minIndex = v;
            }
        }

        return minIndex;
    }
}

public class Program
{
    public static void Main()
    {
        int vertices = 9;
        Graph graph = new Graph(vertices);

        graph.AddEdge(0, 1, 4);
        graph.AddEdge(0, 7, 8);
        graph.AddEdge(1, 2, 8);
        graph.AddEdge(1, 7, 11);
        graph.AddEdge(2, 3, 7);
        graph.AddEdge(2, 8, 2);
        graph.AddEdge(2, 5, 4);
        graph.AddEdge(3, 4, 9);
        graph.AddEdge(3, 5, 14);
        graph.AddEdge(4, 5, 10);
        graph.AddEdge(5, 6, 2);
        graph.AddEdge(6, 7, 1);
        graph.AddEdge(6, 8, 6);
        graph.AddEdge(7, 8, 7);

        int source = 0;
        int[] distances = graph.Dijkstra(source);

        Console.WriteLine("Vertex\tDistance from Source");
        for (int i = 0; i < vertices; i++)
        {
            Console.WriteLine($"{i}\t{distances[i]}");
        }
    }
}

File: src/snippets/graph/dijkstra/dijkstra.lua
-- Dijkstra's Algorithm in Lua

-- Function to find the vertex with the minimum distance value
local function minDistance(dist, sptSet, V)
    local min = math.huge
    local min_index = -1

    for v = 1, V do
        if sptSet[v] == false and dist[v] <= min then
            min = dist[v]
            min_index = v
        end
    end

    return min_index
end

-- Function to implement Dijkstra's algorithm
local function dijkstra(graph, src, V)
    local dist = {}
    local sptSet = {}

    -- Initialize all distances as infinite and sptSet[] as false
    for i = 1, V do
        dist[i] = math.huge
        sptSet[i] = false
    end

    -- Distance of source vertex from itself is always 0
    dist[src] = 0

    -- Find shortest path for all vertices
    for count = 1, V - 1 do
        -- Pick the minimum distance vertex from the set of vertices not yet processed
        local u = minDistance(dist, sptSet, V)

        -- Mark the picked vertex as processed
        sptSet[u] = true

        -- Update dist value of the adjacent vertices of the picked vertex
        for v = 1, V do
            if not sptSet[v] and graph[u][v] ~= 0 and dist[u] ~= math.huge and dist[u] + graph[u][v] < dist[v] then
                dist[v] = dist[u] + graph[u][v]
            end
        end
    end

    return dist
end

-- Example usage
local graph = {
    {0, 10, 0, 0, 0, 0},
    {10, 0, 5, 0, 0, 0},
    {0, 5, 0, 20, 1, 0},
    {0, 0, 20, 0, 2, 2},
    {0, 0, 1, 2, 0, 3},
    {0, 0, 0, 2, 3, 0}
}

local V = 6
local src = 1
local distances = dijkstra(graph, src, V)

print("Vertex Distance from Source")
for i = 1, V do
    print(i, distances[i])
end